activemodel/lib/active_model/attribute_methods.rb

```diff
@@ -203,34 +206,77 @@ def attribute_method_affix(*affixes)
       #   person.nickname_short? # => true
       def alias_attribute(new_name, old_name)
         self.attribute_aliases = attribute_aliases.merge(new_name.to_s => old_name.to_s)
-        ActiveSupport::CodeGenerator.batch(self, __FILE__, __LINE__) do |code_generator|
+        eagerly_generate_alias_attribute_methods(new_name, old_name)
+      end
+
+      def eagerly_generate_alias_attribute_methods(new_name, old_name) # :nodoc:
+        generate_alias_attribute_methods(new_name, old_name)
+      end
+
+      def generate_alias_attribute_methods(new_name, old_name)
+        ActiveSupport::CodeGenerator.batch(generated_attribute_methods, __FILE__, __LINE__) do |code_generator|
           attribute_method_patterns.each do |pattern|
             method_name = pattern.method_name(new_name).to_s
             target_name = pattern.method_name(old_name).to_s
             parameters = pattern.parameters
+            old_name = old_name.to_s
+
+            method_defined_within = method_defined_within?(target_name, self)
+            manually_defined_on_class = method_defined_within && self.instance_method(target_name).owner == self
 
-            mangled_name = target_name
-            unless NAME_COMPILABLE_REGEXP.match?(target_name)
-              mangled_name = "__temp__#{target_name.unpack1("h*")}"
+            if manually_defined_on_class
+              aliased_method_redefined_as_well = method_defined_within?(method_name, self)
+              next if aliased_method_redefined_as_well
+
+              ActiveModel.deprecator.warn(
+                "#{self} model aliases #{old_name} and has a method called #{target_name} already defined. " \
+                "Since Rails 7.2 #{method_name} will not be calling #{target_name} anymore. " \
+                "You may want to additionally define #{method_name} to preserve the current behavior."
+              )
+              deprecated_alias_attribute_definition(code_generator, method_name, target_name, parameters)
+            else
+              define_proxy_call(code_generator, method_name, pattern.proxy_target, parameters, old_name,
+                namespace: :proxy_alias_attribute
+              )
```

Comment:
Ah, sorry, I completely misunderstood that 🤦 I'll change it!

Reply:
We started this journey to support instances where `old_name` cannot be defined (because it's a reserved method name)

Reply:
Right, `alias_method ` won't work for `id` and other reserved names since it will copy the non-attribute method definition 😞 

There is also a smaller concern with `alias_method` - we need to make sure that attribute methods have been generated prior alias attribute methods. Not a big problem but something that will have to be implicitly maintained

Reply:
> Right, `alias_method ` won't work for `id` and other reserved names since it will copy the non-attribute method definition 😞

How so? If you do the `alias_method` inside the `@generated_attribute_methods` module, it will alias the attribute methods. No?

Reply:
It's not that it'll alias the wrong thing, but that for some old_names (which we currently do not support at all), we can't define the original old_name method at all [because it would override an AR builtin]. I believe `id` is special in that regard, because its real/wanted definition arrives after, and overrides, the attribute method.

We can still proper-alias where possible, I'm just noting that there will be exceptions where we need to directly define a method A as an accessor for column B.

Reply:
I see. Yeah, I think that's fine. We can do the aliases last, and then check `@mod.method_defined?(old_name)`.